src/FSharpVSPowerTools.Core/Utils.fs

```diff
@@ -27,60 +48,136 @@ module List =
             | _ -> List.rev acc
         loop [] xs
 
+    /// Fold over the list passing the index and element at that index to a folding function
+    let foldi (folder:'State -> int -> 'T -> 'State) (state:'State) (xs:'T list) =        
+        match xs with 
+        | [] -> state
+        | _ -> 
+            let f = OptimizedClosures.FSharpFunc<_,_,_,_>.Adapt folder
+            let rec loop idx s xs = 
+                match xs with 
+                | [] -> s
+                | h::t -> loop (idx+1) (f.Invoke(s,idx,h)) t
+            loop 0 state xs
+
+
+    let inline groupBy (keyfn:'T->'Key) (list: 'T list) =
+        let dict = Dictionary<'Key,ResizeArray<'T>> (HashIdentity.Structural)
+        // Build the groupings
+        let rec loop list =
+            match list with
+            | v :: t -> 
+                let key = keyfn v
+                let ok,prev = dict.TryGetValue key
+                if ok then
+                    prev.Add v
+                else 
+                    let prev = new ResizeArray<'T> 1
+                    dict.[key] <- prev
+                    prev.Add v
+                loop t
+            | _ -> ()
+        loop list
+        dict  // Return the list-of-lists.
+        |> Seq.map (fun group -> (group.Key, group.Value.ToArray() |> List.ofArray))
```

Comment:
```
groupby seq 0.55ms
groupby arr 0.14ms
Ran 100 times
Ratio:  3.928571429
groupby seq 1.55ms
groupby arr 1.7ms
Ran 100 times
Ratio:  0.9117647059
groupby seq 20.19ms
groupby arr 21.12ms
Ran 100 times
Ratio:  0.9559659091
groupby seq 199.47ms
groupby arr 212.42ms
Ran 100 times
Ratio:  0.9390358723
```
It seemed like we'd be using it within the bounds where using an array internally was faster
[full code](https://gist.github.com/cloudRoutine/ada3b853497749024411)